DataManagementSystem/Client/ConsistencyChecks.py

```diff
@@ -0,0 +1,836 @@
+
+    res = se.exists( lfnDir )
+    if not res['OK']:
+      gLogger.error( "Failed to obtain existance of directories", res['Message'] )
+      return res
+    for directory, error in res['Value']['Failed'].items():
+      gLogger.error( 'Failed to determine existance of directory', '%s %s' % ( directory, error ) )
+    if res['Value']['Failed']:
+      return S_ERROR( 'Failed to determine existance of directory' )
+    directoryExists = res['Value']['Successful']
+    activeDirs = []
+    for directory in sorted( directoryExists ):
+      exists = directoryExists[directory]
+      if exists:
+        activeDirs.append( directory )
+    allFiles = {}
+    while len( activeDirs ) > 0:
+      currentDir = activeDirs[0]
+      res = se.listDirectory( currentDir )
+      activeDirs.remove( currentDir )
+      if not res['OK']:
+        gLogger.error( 'Failed to get directory contents', res['Message'] )
+        return res
+      elif currentDir in res['Value']['Failed']:
+        gLogger.error( 'Failed to get directory contents', '%s %s' % ( currentDir, res['Value']['Failed'][currentDir] ) )
+        return S_ERROR( res['Value']['Failed'][currentDir] )
+      else:
+        dirContents = res['Value']['Successful'][currentDir]
+        activeDirs.extend( se.getLFNFromURL( dirContents['SubDirs'] ).get( 'Value', {} ).get( 'Successful', [] ) )
+        fileURLMetadata = dirContents['Files']
+        fileMetadata = {}
+        res = se.getLFNFromURL( fileURLMetadata )
+        if not res['OK']:
+          gLogger.error( 'Failed to get directory content LFNs', res['Message'] )
+          return res
+
+        for url, error in res['Value']['Failed'].items():
+          gLogger.error( "Failed to get LFN for URL", "%s %s" % ( url, error ) )
+        if res['Value']['Failed']:
+          return S_ERROR( "Failed to get LFNs for PFNs" )
+        urlLfns = res['Value']['Successful']
+        for urlLfn, lfn in urlLfns.items():
+          fileMetadata[lfn] = fileURLMetadata[urlLfn]
+        allFiles.update( fileMetadata )
+
+    zeroSizeFiles = []
+
+    for lfn in sorted( allFiles ):
+      if os.path.basename( lfn ) == 'dirac_directory':
+        allFiles.pop( lfn )
+      else:
+        metadata = allFiles[lfn]
+        if metadata['Size'] == 0:
+          zeroSizeFiles.append( ( lfn, 'deprecatedUrl', storageElement, 'PFNZeroSize' ) )
+    if zeroSizeFiles:
+      self.__reportProblematicReplicas( zeroSizeFiles, storageElement, 'PFNZeroSize' )
+
+    gLogger.info( 'Obtained at total of %s files for directories at %s' % ( len( allFiles ), storageElement ) )
+    return S_OK( allFiles )
+
+  def __getCatalogDirectoryContents( self, lfnDir ):
+    """ Obtain the contents of the supplied directory
+    """
+    gLogger.info( 'Obtaining the catalog contents for %s directories' % len( lfnDir ) )
+
+    activeDirs = lfnDir
+    allFiles = {}
+    while len( activeDirs ) > 0:
+      currentDir = activeDirs[0]
+      res = self.fc.listDirectory( currentDir )
```

Comment:
listDirectory in filecatalog has a recursive option, which avoids you to implement it yourself